Integrate Flash Storage

H1ghBre4k3r opened the pull request:
This set of commits adds support for using the onboard flash of nRF-boards. 

The flash can be utilized by a thin layer on top of the flash with fixed entry size, which manages all actions regarding addition, deletion and reading of entries, and can be initialized for any custom or predefined flash partition. Furthermore, the implementation of a storage-api provides an interface for storing and reading exposure contacts, upon which one can perform actions by using a the iterator-api.

H1ghBre4k3r committed:
Add device tree overlay and ini entry for nrf52840_mdk

H1ghBre4k3r committed:
Add storage initialization

H1ghBre4k3r committed:
Add 0 as success return value for init_storage

H1ghBre4k3r committed:
Enable flash in prj conf

H1ghBre4k3r committed:
Add DS_Store to .gitignore

H1ghBre4k3r committed:
Move period_contact_t to header file

H1ghBre4k3r committed:
Move type declarations in contacts.h

H1ghBre4k3r committed:
Add store_period_contacts

H1ghBre4k3r committed:
get_contacts_for_period

H1ghBre4k3r committed:
Implement api for storage

H1ghBre4k3r committed:
Add storage changes

H1ghBre4k3r committed:
Add records api, move sequence numbers into own file, adjust storage api

H1ghBre4k3r committed:
Add sn_mask to mask a sequence number

H1ghBre4k3r committed:
Correctly increase sequence number and save storage information to flash

H1ghBre4k3r committed:
Rename sequence number functions to fit pattern

H1ghBre4k3r committed:
Handle contact deletion for contact_information

H1ghBre4k3r committed:
Add support for external onboard-storage

H1ghBre4k3r committed:
Increase NVS to 192 sectors, 16.384 byte each

H1ghBre4k3r committed:
Add mutex for entire NVS

H1ghBre4k3r committed:
Automatically increment sequence number

H1ghBre4k3r committed:
Implement ens_fs and move ens stuff to ens directory

H1ghBre4k3r committed:
Move sequence-number utils to ens folder aswell

H1ghBre4k3r committed:
Rename storage partition

H1ghBre4k3r committed:
Use NVS to store storage information

H1ghBre4k3r committed:
Init ens_fs in storage.c

H1ghBre4k3r committed:
Store contacts in ens_fs

H1ghBre4k3r committed:
Store CRC and 'not-deleted' flag in last byte of each entry

H1ghBre4k3r committed:
Delete page before incrementing sn

H1ghBre4k3r committed:
Implement correct read using crc and deleted flag and implement deletion

H1ghBre4k3r committed:
Use record_t for storing contacts

H1ghBre4k3r committed:
Implement error codes and use them throughout ens_fs

H1ghBre4k3r committed:
Remove unneeded contacts code and cleanup gatt-service.h

H1ghBre4k3r committed:
Fix incompatible pointer in ens_fs and comment unneeded variables out

H1ghBre4k3r committed:
Implement ens_records_iterator_init_timerange via binary search

H1ghBre4k3r committed:
Use binary search in init_timerange

H1ghBre4k3r committed:
Format ens files and add .clang-format file

H1ghBre4k3r committed:
Add locks for the storage

prathje commented on for_future_use.txt:
```diff
@@ -0,0 +1,219 @@
+
```
This file should probably not be checked-in?

prathje commented on platformio.ini:
```diff
@@ -28,6 +28,23 @@ lib_deps =
 lib_ignore =
     mbedtls
 
+[env:nrf52840_mdk]
```
Since this config seems to match the other nrf config. Could you merge them together using PIO env extension?

prathje commented on src/covid.c:
```diff
@@ -74,7 +75,16 @@ static void scan_cb(const bt_addr_le_t *addr, int8_t rssi, uint8_t adv_type, str
+                    memcpy(&contact.rolling_proximity_identifier, &rx_adv->rolling_proximity_identifier, sizeof(contact.rolling_proximity_identifier));
+                    memcpy(&contact.timestamp, &timestamp, sizeof(contact.timestamp));
+                    int rc = add_contact(&contact);
+                    printk("Contact stored (err %d)\n", rc);
```
Does this implicitly handle the case when the storage is full?

H1ghBre4k3r replied:
Yes, it does. If the storage is full, a page at the start of the flash area will be erased and we start storing contacts at the start of this page.

prathje commented on src/covid.c:
```diff
@@ -74,7 +75,16 @@ static void scan_cb(const bt_addr_le_t *addr, int8_t rssi, uint8_t adv_type, str
+                    memcpy(&contact.timestamp, &timestamp, sizeof(contact.timestamp));
+                    int rc = add_contact(&contact);
+                    printk("Contact stored (err %d)\n", rc);
+					// check_add_contact(k_uptime_get() / 1000, &rx_adv->rolling_proximity_identifier, &rx_adv->associated_encrypted_metadata, rssi);
```
```suggestion
```

prathje commented on src/covid.c:
```diff
@@ -101,14 +111,14 @@ static void test_against_fixtures(void)
 	ENIntervalNumber intervalNumber = 2642976;
 	ENPeriodKey periodKey = {.b = {0x75, 0xc7, 0x34, 0xc6, 0xdd, 0x1a, 0x78, 0x2d, 0xe7, 0xa9, 0x65, 0xda, 0x5e, 0xb9, 0x31, 0x25}};
-	unsigned char metadata[4] = {0x40, 0x08, 0x00, 0x00};
+	// unsigned char metadata[4] = {0x40, 0x08, 0x00, 0x00};
```
Why were those lines removed? Aren't they used in the fixture test?

prathje commented on src/covid.c:
```diff
@@ -312,10 +322,11 @@ static void check_keys(struct k_work *work)
+		// if (!init)
+		// {
+		// 	key_change(current_period_index);
+		// }
```
I think that we can safely delete this part now?
```suggestion
```

prathje commented on src/ens/ens_fs.c:
```diff
@@ -0,0 +1,103 @@
+#include "ens_error.h"
+#include "ens_fs.h"
+
+#define SEED 42
```
Is this SEED required? For what reason?
If it is still required, please introduce a KConfig Variable

prathje replied:
Used as a seed for CRC calculations

prathje commented on src/ens/ens_fs.c:
```diff
@@ -0,0 +1,103 @@
+
+#define SEED 42
+
+int ens_fs_init(ens_fs_t* fs, uint8_t id, uint64_t entry_size) {
```
The entry_size here is not the real size of an entry as we include another byte for CRC and deletion flag?
This seems a bit confusing to me?

prathje replied:
Please also rename the id to flash_id
```suggestion
int ens_fs_init(ens_fs_t* fs, uint8_t flash_id, uint64_t entry_size) {
```

prathje commented on src/ens/ens_fs.c:
```diff
@@ -0,0 +1,103 @@
+    return 0;
+}
+
+int ens_fs_read(ens_fs_t* fs, uint64_t id, void* dist) {
```
```suggestion
int ens_fs_read(ens_fs_t* fs, uint64_t id, void* dest) {
```

prathje commented on src/ens/ens_fs.c:
```diff
@@ -0,0 +1,103 @@
+
+    // check, if checksum and not-deleted flag are as expected
+    uint8_t* obj = dist;
+    uint8_t metadata = crc7_be(SEED, obj, fs->entry_size - 1) | 1;
```
Please check the deletion separately, i.e. introduce another return value for when the entry was deleted.
Use bit masks and bit shifts to check for deletion and also for the metadata.

prathje commented on src/ens/ens_fs.c:
```diff
@@ -0,0 +1,103 @@
+int ens_fs_write(ens_fs_t* fs, uint64_t id, void* data) {
+    // set CRC and not-deleted-flag
+    uint8_t* obj = data;
+    obj[fs->entry_size - 1] = crc7_be(SEED, obj, fs->entry_size - 1) | 1;
```
This is a good example why we should separate the entry_size from the saved user data: we should not change the underlying data this way.

prathje replied:
For a quick fix: We could specify a maximum supported length and allocate that on the stack?

prathje commented on src/ens/ens_fs.c:
```diff
@@ -0,0 +1,103 @@
+    return 0;
+}
+
+int ens_fs_page_erase(ens_fs_t* fs, uint64_t offset, uint64_t sector_count) {
```
Please also use the id as the way to determine the offset.

prathje commented on src/ens/ens_fs.h:
```diff
@@ -0,0 +1,90 @@
+    /**
+     * Lock for this fs.
+     */
+    // struct k_mutex* lock;
```
Do we need this lock or not?

prathje commented on src/ens/records.c:
```diff
@@ -0,0 +1,164 @@
+                                    record_sequence_number_t* opt_end) {
+    iterator->sn_next = opt_start ? *opt_start : get_oldest_sequence_number();
+    iterator->sn_end = opt_end ? *opt_end : get_latest_sequence_number();
+    if (get_num_contacts() == 0) {
```
What if the oldest, last or amount of contact changes due to another thread deleting or writing records?

H1ghBre4k3r replied:
What could we do against this? Locking a mutex for the entire iteration would block other threads from changing our stored records but would also limit the possibility of multithreading. 
Alternatively, we could introduce an update-function, which allows us to update the start and end of our iterator depending on the oldest and latest sequence number.

prathje commented on src/ens/records.c:
```diff
@@ -0,0 +1,164 @@
+        return rc;
+    }
+    // if starting timestamp lies in our bounds, perform binary search
+    if (start_rec.timestamp < *ts_start) {
```
Caution: start and end timestamps are optional, this is why they are defined as pointers.
In case the start is not defined: just use oldest_sn, binary search for start otherwise
In case the end is not defined: use latest_sn for end_sn, binary search for end otherwise

prathje replied:
Also: There is no need to load contacts in this function. -> Maybe move the oldest_sn and latest_sn to the binary_search method? -> this saves some arguments ;)

prathje replied:
Please also consider to keep track of the respective timestamp of the oldest and the newest record (but this could be an optimization for later).

prathje commented on src/ens/records.c:
```diff
@@ -0,0 +1,164 @@
+    }
+    // if starting timestamp lies in our bounds, perform binary search
+    if (start_rec.timestamp < *ts_start) {
+        rc = find_record_via_binary_search(&start_rec, *ts_start, oldest_sn, latest_sn);
```
The binary search does not need to return the actual record. Instead we are solely interested in the sequence number. Even if we did not find the exact record. The oldest or latest entry might still fall into the relevant timespan!

prathje commented on src/ens/records.c:
```diff
@@ -0,0 +1,164 @@
+        }
+    }
+
+    iterator->finished = false;
```
Please consider calling the existing iterator init function here as we now both start and end sn.

prathje commented on src/ens/records.c:
```diff
@@ -0,0 +1,164 @@
+    record_t* next = NULL;
+
+    // What is this?
+    while (next == NULL) {
```
This will block in case we reached the last record but could not load it.
We also need to check that the iter->finished == false.
```suggestion
    while (next == NULL && !iter->finished) {
```

prathje commented on src/ens/records.c:
```diff
@@ -0,0 +1,164 @@
+record_t* ens_records_iterator_next(record_iterator_t* iter) {
+    if (iter->finished) {
+        return NULL;
+    }
```
I think that we need to integrate this into the while loop (see below)
```suggestion
```

prathje commented on src/ens/records.c:
```diff
@@ -0,0 +1,164 @@
+
+        if (!res) {
+            next = &iter->current;
+            memcpy(&next->associated_encrypted_metadata, &contact.associated_encrypted_metadata,
```
Consider writing a simple copy function for records?

prathje commented on src/ens/records.c:
```diff
@@ -0,0 +1,164 @@
+}
+
+int ens_record_iterator_clear(record_iterator_t* iter) {
+    // clear all relevant fields in the iterator
```
Set iter->sn_next and iter->sn_end to zero, too?

prathje commented on src/ens/records.c:
```diff
@@ -0,0 +1,164 @@
+    iterator->sn_end = opt_end ? *opt_end : get_latest_sequence_number();
+    if (get_num_contacts() == 0) {
+        iterator->finished = true;  // no contacts -> no iteration :)
+    }
```
We need to set finished to false in case the underlying memory has been previously used as an iterator.
```suggestion
    }
    else {
        iterator->finished = false;
    
    }
```

prathje commented on src/ens/sequencenumber.h:
```diff
@@ -0,0 +1,36 @@
+ * @param sn sequence number to mask
+ * @return masked sequence number
+ */
+record_sequence_number_t sn_mask(record_sequence_number_t sn);
```
Yes, please introduce that as a macro. This should also be defined in the source file as this is an internal operation.

prathje commented on src/ens/storage.c:
```diff
@@ -0,0 +1,175 @@
+#define SEC_COUNT 8U
+
+#define STORED_CONTACTS_INFO_ID 0
+#define MAX_CONTACTS 65536
```
MAX_CONTACTS should be removed. The number of maximum storable contacts should be directly depend on the given flash size!

prathje replied:
Or the flash size depends on the MAX_CONTACTS (but then please introduce this as a Kconfig variable).

prathje commented on src/ens/storage.c:
```diff
@@ -0,0 +1,175 @@
+#include "storage.h"
+
+// Maybe use this as param for init function
+#define SEC_COUNT 8U
```
This might either be a config value or information that we should infer from the flash info?

prathje commented on src/ens/storage.c:
```diff
@@ -0,0 +1,175 @@
+    int rc = nvs_read(&info_fs, STORED_CONTACTS_INFO_ID, &contact_information, size);
+
+    // Check, if read what we wanted
+    if (rc != size) {
+        // Write our initial data to storage
+        rc = nvs_write(&info_fs, STORED_CONTACTS_INFO_ID, &contact_information, size);
+        if (rc <= 0) {
+            return rc;
```
This seems a bit odd? Why would we write the info in case the read was not successful?

H1ghBre4k3r replied:
On first startup, our flash is completely empty. In this case, our read is not successful and writing the initial configuration (0 stored contacts) brings our flash into a clean state.

prathje commented on src/ens/storage.c:
```diff
@@ -0,0 +1,175 @@
+        contact_information.oldest_contact = sn_increment(contact_information.oldest_contact);
+    } else {
+        contact_information.count++;
+    }
```
I don't think that we need to actually check if we reached the maximum amount of contacts.
In fact, I think that this function might not need to save the storage information at all. 
We could implement this functionality entirely in the add_contact function as we only need a new sequence number when adding new contacts (see below) - this would remove the need for this internal function entirely?

prathje commented on src/ens/storage.c:
```diff
@@ -0,0 +1,175 @@
+        return rc;
+    }
+    info_fs.sector_size = info.size;
+    info_fs.sector_count = SEC_COUNT;
```
Can we infer this from the flash device?

prathje commented on src/ens/storage.c:
```diff
@@ -0,0 +1,175 @@
+    // Load the current storage information
+    rc = load_storage_information();
+    if (rc) {
+        printk("Cannot load storage information (err %d)\n", rc);
```
```suggestion
        printk("Cannot load storage information (err %d)\n", rc);
        // TODO: Try to recover entries or bring the flash into a clean state?
```

prathje commented on src/ens/storage.c:
```diff
@@ -0,0 +1,175 @@
+    printk("Currently %d contacts stored!\n", contact_information.count);
+    printk("Space available: %d\n", FLASH_AREA_SIZE(storage));
+
+    // TODO lome: change size to sizeof(contact_struct)
```
Yes, please change it to that size :)

prathje commented on src/ens/storage.c:
```diff
@@ -0,0 +1,175 @@
+
+int load_contact(record_t* dest, record_sequence_number_t sn) {
+    storage_id_t id = convert_sn_to_storage_id(sn);
+    // int rc = nvs_read(&info_fs, id, dest, sizeof(*dest));
```
```suggestion
```

prathje commented on src/ens/storage.c:
```diff
@@ -0,0 +1,175 @@
+    return 0;
+}
+
+int add_contact(record_t* src) {
```
Please be consistent in the naming of contact and record. Which one do you prefer?

prathje commented on src/ens/storage.c:
```diff
@@ -0,0 +1,175 @@
+    if (((potential_next_id * ens_fs.entry_size) % ens_fs.sector_size) == 0) {
+        // If we are at start of a page, we need to erase it first
+        ens_fs_page_erase(&ens_fs, potential_next_id, 1);
+    }
```
If we delete a page, we automatically lose a few records -> we thus need to adapt the oldest sn (and the respective count).

prathje commented on src/ens/storage.c:
```diff
@@ -0,0 +1,175 @@
+    src->sn = curr_sn;
+    storage_id_t id = convert_sn_to_storage_id(curr_sn);
+
+    return ens_fs_write(&ens_fs, id, src);
```
I guess we should save the info in this function (as we need to change it for the deletion anyway?).

prathje commented on src/ens/storage.c:
```diff
@@ -0,0 +1,175 @@
+    if (!rc) {
+        k_mutex_lock(&info_fs_lock, K_FOREVER);
+        if (sn_equal(sn, get_latest_sequence_number())) {
+            contact_information.count--;
```
Why can we safely lower the count if this is the latest sequence number but not in between?
This  count decrement would essentially allow us to use a sequence number twice, correct?
But as we do not support writing twice to the same sn, this would result in an invalid entry.

prathje commented on src/ens/storage.c:
```diff
@@ -0,0 +1,175 @@
+    return rc;
+}
+
+// TODO lome: do we need lock here aswell?
```
I do not think we need to lock this for every use case but instead every function that manages/modifies the contact_information struct. For instance, an iterator might read outdated or even an invalid latest_sequence_number. Worst case would be that an iterator loops through invalid data. (Hold on a minute, this essentially means that due to e.g. a page erase and init with new and thus higher sequence numbers the iterator might loop through wrong entries, right? -> so we might actually need the saved sequence number to check that we loaded the correct entry!?!) This of course will not prevent the same issue resulting from the sequence number wrap-around :) (but that should be a lot less infrequent and technically you would read the correct sequence number?). Alternative: Check the sn before reading and ensure that the sn lies within our bounds of available sequence numbers? (not yet what would need to be locked for that though? -> comparing sequence numbers seems faster atm).

prathje commented on src/ens/storage.c:
```diff
@@ -0,0 +1,175 @@
+}
+
+uint32_t get_num_contacts() {
+    return contact_information.count;
```
We might want to distinguish between the real amount of contacts and the amount of used sequence numbers, maybe?

prathje commented on src/ens/storage.h:
```diff
@@ -0,0 +1,72 @@
+
+typedef uint16_t storage_id_t;
+
+// TODO lome: own datatype for storage
```
I think that the storage should just work with the existing record struct.

prathje commented on src/ens/storage.h:
```diff
@@ -0,0 +1,72 @@
+
+typedef struct stored_contacts_information {
+    record_sequence_number_t oldest_contact;
+    record_sequence_number_t count;
```
Count should not be a sequence number! (32 bits should be enough I guess)
```suggestion
    uint32_t count;
```

prathje commented on src/ens/storage.h:
```diff
@@ -0,0 +1,72 @@
+record_sequence_number_t get_oldest_sequence_number();
+
+/**
+ * TODO: How to handle if none is available?
```
This is handled excellently with count.
```suggestion
```

prathje commented on zephyr/prj.conf:
```diff
@@ -15,3 +15,27 @@ CONFIG_MBEDTLS_HEAP_SIZE=16384
+
+CONFIG_LOG=y
+
+CONFIG_FLASH=y
+CONFIG_FLASH_MAP=y
+CONFIG_FLASH_PAGE_LAYOUT=y
```
CONFIG_FLASH and FLASH_PAGE_LAYOUT are already defined above.

prathje commented on zephyr/prj.conf:
```diff
@@ -15,3 +15,27 @@ CONFIG_MBEDTLS_HEAP_SIZE=16384
+CONFIG_FLASH_MAP=y
+CONFIG_FLASH_PAGE_LAYOUT=y
+
+# Need this when storage is on MX25R64
+CONFIG_NORDIC_QSPI_NOR=y
+CONFIG_NORDIC_QSPI_NOR_FLASH_LAYOUT_PAGE_SIZE=4096
```
If this is only needed for MX25R64, please consider to move this to the PIO.ini file or a separate board conf file.

prathje reviewed (changes requested):
Hey @H1ghBre4k3r,
Thank you really much for your work!
I went through your changes and commented on what I think we should improve.
Please check what changes you would want to make yourself and which I should do myself.

Best Regards
Patrick

H1ghBre4k3r committed:
Remove unneeded text file

H1ghBre4k3r committed:
Use environment extension for platformio.ini

H1ghBre4k3r committed:
Apply suggestions for covid.c

H1ghBre4k3r committed:
Apply suggestions for ens_fs

H1ghBre4k3r committed:
Use lock for ens_fs

H1ghBre4k3r committed:
Records: In init_timerange, use init_range after finding start and end sn

H1ghBre4k3r committed:
Records: In iterator_next, combine if-guard with while-loop

H1ghBre4k3r committed:
Records: In iterator_next, use memcpy to next contact

H1ghBre4k3r committed:
Records: In iterator_clear, also set sequence numbers to 0

H1ghBre4k3r committed:
Records: In init_range, set finished to false, if there are contacts

H1ghBre4k3r committed:
Sequence Numbers: Introduce GET_MASKED_SN macro

H1ghBre4k3r committed:
Storage: Remove comment in load_contact

H1ghBre4k3r committed:
Use record instead of contact as naming convention

H1ghBre4k3r committed:
Storage: Adjust record_information correctly when erasing pages and deleting contacts

H1ghBre4k3r committed:
Storage: Use correct datatypes for storage information

H1ghBre4k3r committed:
Prj-Conf: Remove already defined config parameters

H1ghBre4k3r committed:
Storage: Add 3 byte for metadata to record_t and use it in initialization for ens_fs

prathje commented on src/ens/ens_fs.c:
```diff
@@ -0,0 +1,144 @@
+    }
+
+end:
+    k_mutex_unlock(&fs->ens_fs_lock);
```
Please unlock as soon as possible to lower the impact of synchronization.

prathje commented on src/ens/ens_fs.h:
```diff
@@ -0,0 +1,90 @@
+     * Lock for this fs.
+     */
+    struct k_mutex ens_fs_lock;
+} ens_fs_t;
```
Add a write_buffer entry that is allocated from the heap (which has the size entry_size+1) or 2^(ceil(log(entry_size+1))

prathje commented on src/ens/ens_fs.c:
```diff
@@ -0,0 +1,144 @@
+    }
+    fs->sector_size = info.size;
+
+    k_mutex_init(&fs->ens_fs_lock);
```
Allocate internal entry buffer.

prathje replied:
Calculate internal entry_size based on the given entry_size. (internal = 2^(ceil(log(entry_size+1))

prathje replied:
We need the internal_entry_size and entry_size.

prathje commented on src/ens/records.c:
```diff
@@ -0,0 +1,154 @@
+    }
+
+    do {
+        // calculate the contact in the middle between start and end and load it
```
We will also need to select the first entry for the start or the last entry for the end.

prathje commented on src/ens/sequencenumber.h:
```diff
@@ -0,0 +1,32 @@
+/**
+ * Mask a given sequence number to get rid of MSB.
+ */
+#define GET_MASKED_SN(x) (x & SN_MASK)
```
Please move this to the source file. And add another function to handle incrementing with a positive amount of steps.

prathje commented on src/ens/sequencenumber.h:
```diff
@@ -0,0 +1,32 @@
+#include <stdint.h>
+
+// mask for sequence numbers (2^24 - 1)
+#define SN_MASK 16777215
```
```suggestion
#define SN_MASK 0xffffff
```

prathje replied:
```suggestion
#define SN_MASK ((1<<24)-1)
```

prathje commented on src/ens/ens_fs.c:
```diff
@@ -0,0 +1,144 @@
+    }
+
+    // check, if entry size is multiple of flash_area_align
+    if ((entry_size % flash_area_align(fs->area)) != 0) {
```
Please also check that internal_entry_size is less than the page size.
Also calculate the amount of entries per Page and in total.

prathje commented on src/ens/ens_fs.c:
```diff
@@ -0,0 +1,144 @@
+    k_mutex_lock(&fs->ens_fs_lock, K_FOREVER);
+
+    // calculate the next page start before (or at) the given entry_id
+    uint64_t start = (entry_id - entry_id % fs->sector_size) * fs->entry_size;
```
This should use the internal_entry_size and not the entry_id to calculate the offset.

prathje commented on src/ens/ens_fs.c:
```diff
@@ -0,0 +1,144 @@
+    return rc;
+}
+
+int ens_fs_page_erase(ens_fs_t* fs, uint64_t entry_id, uint64_t sector_count) {
```
```suggestion
// Returns the amount of theoretically deleted entries (e.g. entries, entry_id up to entry_id+ret could be deleted)
uint64_t ens_fs_make_space(ens_fs_t* fs, uint64_t entry_id) {
```

prathje replied:
The function should return an error if entry_id is not at the start of the page.

prathje commented on src/ens/ens_fs.c:
```diff
@@ -0,0 +1,144 @@
+    obj[fs->entry_size - 1] = crc7_be(SEED, obj, fs->entry_size - 1) | 1;
+
+    uint64_t offset = id * fs->entry_size;
+    if (flash_area_write(fs->area, offset, data, fs->entry_size)) {
```
Before we write, we need to check that at this position the flash area is not used!

prathje replied:
If it is unitialized (a zero in one of the positions of the loaded buffer) then we return occupied or so.

prathje commented on src/ens/storage.c:
```diff
@@ -0,0 +1,185 @@
+    // Check, if next sn would be at start of page
+    record_sequence_number_t potential_next_sn = sn_increment(get_latest_sequence_number());
+    storage_id_t potential_next_id = convert_sn_to_storage_id(potential_next_sn);
+    if (((potential_next_id * ens_fs.entry_size) % ens_fs.sector_size) == 0) {
```
We need to rework this: We simply try to write all the time (which might fail due to space constraints or invalid flash memory). If writing failed, we first try to make_space in the ens system (which also fails if we are in the middle of a page). If it succeeds, we simply try to write again. If clearing also fails, we increase, delete the occupied entry (to prevent future invalid data), increase the used sequence number, and try everything again.

prathje commented on src/ens/records.c:
```diff
@@ -0,0 +1,154 @@
+ * @param start lower bound for the binary search
+ * @param end upper bound for the binary search
+ */
+int find_record_via_binary_search(record_t* record,
```
BTW: Some entries could be deleted or invalid. We also need to handle those cases, i.e. loading the entry "to the left/right".

prathje replied:
Caution: We need to make sure that this does not deadlock due to missing entries.

prathje replied:
Check if the oldest ts is newer than the latest -> return error that needs to be handled by the calling function (or the user).

prathje commented on src/ens/sequencenumber.h:
```diff
@@ -0,0 +1,32 @@
+ * @return the incremented sequenced number
+ */
+record_sequence_number_t sn_increment(record_sequence_number_t sn);
+#endif
```
Also add a function to determine the middle sn that handles the case of sequence numbers that wrapped around.

prathje commented on src/ens/records.c:
```diff
@@ -0,0 +1,154 @@
+    return 0;
+}
+
+int ens_records_iterator_init_timerange(record_iterator_t* iterator, uint32_t* ts_start, uint32_t* ts_end) {
```
```suggestion
// TODO: This iterator does neither check if the sequence numbers wrapped around while iteration. As a result, first results could have later timestamps than following entries
int ens_records_iterator_init_timerange(record_iterator_t* iterator, uint32_t* ts_start, uint32_t* ts_end) {
```

H1ghBre4k3r committed:
Fix multiple results of the PR:

- cleanup of sequence number utilites
- dynamic calculation of nvs sectors for fs information
- usage of buffer in ens_fs
- check for max entry size (page size)
- usage of interal entry size, which is always the next larger power of 2 to a given entry size

H1ghBre4k3r committed:
ENS_FS: Only copy memory if read entry is valid

H1ghBre4k3r committed:
ENS_FS: Fix all wrong usages of fs->entry_size

H1ghBre4k3r committed:
Config: Move parameters for MX25R64 into platformio.ini file

H1ghBre4k3r committed:
Add TODOs for refactoring according to PR review and refactor ens_fs_write to return error, if entry is not only ones

H1ghBre4k3r committed:
ENS_FS: Fix wrong bitmask

H1ghBre4k3r committed:
Config: Introduce CONFIG_ENS_MAX_CONTACTS as possibility to configure the max amount of contacts stored in flash

H1ghBre4k3r committed:
SN: Finish impl of sn_get_middle_sn to cover case of older > newer

H1ghBre4k3r committed:
Records: Rewrite interator_init_timerange and binary_search to take care of invalid entries etc.

prathje commented on src/ens/sequencenumber.c:
```diff
@@ -0,0 +1,28 @@
+}
+
+record_sequence_number_t sn_get_middle_sn(record_sequence_number_t older, record_sequence_number_t newer) {
+    if (older < newer) {
```
```suggestion
    if (older <= newer) {
```

prathje replied:
```suggestion
    if (older <= newer) {
```

prathje committed:
Update src/ens/sequencenumber.c

Support equal values of older and newer

prathje reviewed (commented):
WIP
